design-proposals/klip-3-serialization-of-single-fields.md

```diff
@@ -0,0 +1,1076 @@
+ |                         | defaulting to ``true```, is used.                                                                      |
+ |                         |                                                                                                        |
+ |                         | Note: this setting has no effect on formats e.g. ``DELIMITED``, that do not support some kind of outer |
+ |                         | record or object.                                                                                      |
+ +-------------------------+--------------------------------------------------------------------------------------------------------+
+ | WRAP_SINGLE_FIELDS      | Shorthand for setting both ``WRAP_SINGLE_KEY`` and ``WRAP_SINGLE_VALUE`` to the same value.            |
+ +-------------------------+--------------------------------------------------------------------------------------------------------+
+ ```
+
+* A new ``serialization.rst`` will be added to the developer guide:
+
+```rst
+_ksql_serialization:
+
+KSQL Serialization
+==================
+
+.. _ksql_formats
+
+=======
+Formats
+=======
+
+KSQL currently supports three serialization formats:
+
+*. ``DELIMITED`` supports comma separated values. See :ref:`delimited_format` below.
+*. ``JSON`` supports JSON values. See :ref:`json_format` below.
+*. ``AVRO`` supports AVRO serialized values. See :ref:`avro_format` below.
+
+
+.. _delimited_format
+
+---------
+DELIMITED
+---------
+
+The ``DELIMITED`` format supports comma separated values.
+
+The serialized object should be a Kafka-serialized string, which will be split into columns.
+
+For example, given a KSQL statement such as:
+
+.. code:: sql
+
+    CREATE STREAM x (ID BIGINT, NAME STRING, AGE INT) WITH (VALUE_FORMAT='DELIMITED', ...);
+
+KSQL splits a value of ``120, bob, 49`` into the three fields with ``ID`` of ``120``,
+``NAME`` of ``bob`` and ``AGE`` of ``49``.
+
+This data format supports all KSQL :ref:`data types <data-types>` except for container types, 
+i.e. ``ARRAY``, ``MAP`` and ``STRUCT``.
+
+.. _json_format
+
+----
+JSON
+----
+
+The ``JSON`` format supports JSON values.
+
+The JSON format supports all of KSQL's ref:`data types <data-types>`. As JSON does not itself
+support a map type, KSQL serializes ``MAP``s as JSON objects.  Because of this, the JSON format
+can only support ``MAP`` objects that have ``STRING`` keys.
+
+The serialized object should be a Kafka-serialized string containing a valid JSON value. The format
+supports JSON objects and top-level primitives, arrays and maps. See below for more info.
+
+JSON Objects
+------------
+
+Values that are JSON objects are probably the most common.
+
+For example, given a KSQL statement such as:
+
+.. code:: sql
+
+    CREATE STREAM x (ID BIGINT, NAME STRING, AGE INT) WITH (VALUE_FORMAT='JSON', ...);
+
+And a JSON value of:
+
+.. code:: json
+
+       {
+         "id": 120,
+         "name": "bob",
+         "age": "49"
+       }
```

Comment:
Unrelated to this KLIP but since you've drafted this wonderful topic on KSQL serialization (:D), can we add a note calling out the fact that field names are case-insensitive? I've seen a lot of confusion around this, e.g., users thinking they have to quote their field names in the `CREATE STREAM` statement and then hitting issues around our currently-incomplete support for quoted identifiers.

Reply:
Let's not call this out in the docs just yet as our handling is NOT good.  The docs are currently misleading. e.g. 

> If the name of a column in your source topic is one of the reserved words in KSQL you can use back quotes to define the column.

As this will only work some of the time!  It'll work for JSON if you make the quoted property name upper cased. But won't work for Avro unless your field name is actually upper cased.

@almog is looking to resolve this soon and can add docs then.